app/records/utils.py

```diff
@@ -1,30 +1,51 @@
 import logging
-from typing import Any, Dict, Optional
+import re
+from typing import Any, Dict
 
 from django.urls import NoReverseMatch, reverse
 from pyquery import PyQuery as pq
 
+logger = logging.getLogger(__name__)
 
-class ValueExtractionError(Exception):
-    pass
 
+def format_link(link_html: str, inc_msg: str = "") -> Dict[str, str]:
+    """
+    Extracts iaid and text from a link HTML string, e.g. "<a href="C5789">DEFE 31</a>"
+    and returns as dict in the format: `{"id":"C5789", "href": "/catalogue/id/C5789/", "text":"DEFE 31"}
+
+    inc_msg includes message with logger if sepcified
+    Ex:inc_msg <method_name>:Record(<id):"
+    """
+    document = pq(link_html)
+    iaid = document.attr("href")
+    try:
+        href = reverse("details-page-machine-readable", kwargs={"id": iaid})
+    except NoReverseMatch:
+        href = ""
+        # warning for partially valid data
+        logger.warning(
+            f"{inc_msg}format_link:No reverse match for details-page-machine-readable with iaid={iaid}"
+        )
+    return {"id": iaid or "", "href": href, "text": document.text()}
 
-logger = logging.getLogger(__name__)
 
-NOT_PROVIDED = "__np__"
+def format_extref_links(html: str) -> str:
+    html = re.sub(
+        r'<a class="extref" href="([0-9a-zA-Z]+)"',
+        lambda m: f'<a href="{reverse("details-page-machine-readable", kwargs={"id": m.group(1)})}"',
+        html,
+    )
+    return html
```

Comment:
I think for individual Internal links there should be some logging mechanism as Warning if they do not comply with the RegEx format.  Also perhaps a provision for handling refs instead of id's.

Reply:
If they don't match the format we are trying to capture then they are not IDs that we are looking to redirect to a record details page.

Reply:
As I mentioned, its looks to be a data issue
Re
> The data is bad Response id shoud be 6556c5f5-0f02-4d03-9b5c-8ec7973794b7 and not c5f5-0f02-4d03-9b5c-8ec7973794b7

So, there no need to update these
 https://github.com/nationalarchives/ds-search/blob/13c6ad4c1a6d3dbfa9e6ad237e5cb0c767662e15/app/records/converters.py#L7-L9

And this
https://github.com/nationalarchives/ds-search/blob/13c6ad4c1a6d3dbfa9e6ad237e5cb0c767662e15/test/records/test_id_formats.py#L13-L15